java-util/src/main/java/io/druid/java/util/common/granularity/Granularity.java

```diff
@@ -0,0 +1,538 @@
+        case HOUR:
+          return new PeriodGranularity(new Period("PT1H"), origin, tz);
+        case SIX_HOUR:
+          return new PeriodGranularity(new Period("PT6H"), origin, tz);
+        case DAY:
+          return new PeriodGranularity(new Period("P1D"), origin, tz);
+        case WEEK:
+          return new PeriodGranularity(new Period("P1W"), origin, tz);
+        case MONTH:
+          return new PeriodGranularity(new Period("P1M"), origin, tz);
+        case QUARTER:
+          return new PeriodGranularity(new Period("P3M"), origin, tz);
+        case YEAR:
+          return new PeriodGranularity(new Period("P1Y"), origin, tz);
+        case ALL:
+          return new AllGranularity();
+        case NONE:
+          return new NoneGranularity();
+        default:
+          throw new IAE("[%s] granularity not supported with strings. Try with Period instead", str);
+      }
+    }
+
+    // Note: This is only an estimate based on the values in period.
+    // This will not work for complicated periods that represent say 1 year 1 day
+    public static GranularityType fromPeriod(Period period)
+    {
+      int[] vals = period.getValues();
+      BitSet bs = new BitSet();
+      for (int i = 0; i < vals.length; i++) {
+        if (vals[i] != 0) {
+          bs.set(i);
+        }
+      }
+
+      if (bs.cardinality() == 0 || bs.cardinality() > 1) {
+        throw new IAE("Granularity is not supported. [%s]", period);
+      } else {
+        final int index = bs.nextSetBit(0);
+
+        if (index == 0) {
+          return GranularityType.YEAR;
+        }
+        if (index == 1) {
+          if (vals[index] == 4) {
+            return GranularityType.QUARTER;
+          }
+          else if (vals[index] == 1) {
+            return GranularityType.MONTH;
+          }
+        }
+        if (index == 2) {
+          return GranularityType.WEEK;
+        }
+        if (index == 3) {
+          return GranularityType.DAY;
+        }
+        if (index == 4) {
+          if (vals[index] == 6) {
+            return GranularityType.SIX_HOUR;
+          }
+          else if (vals[index] == 1) {
+            return GranularityType.HOUR;
+          }
+        }
+        if (index == 5) {
+          if (vals[index] == 30) {
+            return GranularityType.THIRTY_MINUTE;
+          }
+          else if (vals[index] == 15) {
+            return GranularityType.FIFTEEN_MINUTE;
+          }
+          else if (vals[index] == 10) {
+            return GranularityType.TEN_MINUTE;
+          }
+          else if (vals[index] == 5) {
+            return GranularityType.FIVE_MINUTE;
+          }
+          else if (vals[index] == 1) {
+            return GranularityType.MINUTE;
+          }
+        }
+        if (index == 6) {
+          return GranularityType.SECOND;
+        }
+      }
+
+      throw new IAE("Granularity is not supported. [%s]", period);
+    }
+
+    static DateTime getDateTime(GranularityType gran, Integer[] vals)
```

Comment:
This method is used in the code path where the call site give you the filePath and expects DateTime. [GranularUnprocessedPathSpec](https://github.com/druid-io/druid/pull/3850/files/27679a25f580b914e180314b6d60aac2ce7c2f43#diff-632d3638579cd432dce6307939579389R82) calls [PeriodGranularity.toDate()](https://github.com/druid-io/druid/pull/3850/files/27679a25f580b914e180314b6d60aac2ce7c2f43#diff-7ab54a218fbe531e106899fdeb21c306R131)